RFC: For AD9144, honour the JESD204 crossbar settings.

gsmecher opened the pull request:
Current AD9144 code provides a place in the ad9144_init_param struct for
this field, but setup code ignores it. I recommend either honouring the
field, or removing the placeholder. I have tested this code lightly on a
FMCDAQ2 setup.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/analogdevicesinc/no-OS?pullRequest=138) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/analogdevicesinc/no-OS?pullRequest=138) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/analogdevicesinc/no-OS?pullRequest=138) it.</sub>

dbogdan commented:
Hi,
Your fix looks right. We will do a test on our side as well and we will merge it.
Thanks,
Dragos

gsmecher commented:
Hi Dragos,

On 2019-02-01 4:14 a.m., Dragos Bogdan wrote:
> Hi,
> Your fix looks right. We will do a test on our side as well and we will
> merge it.

Great -- in that case, I'd suggest you also initialize all 8 lanes'
worth of crossbar in fmcdaq2/fmcdaq2.c. Otherwise, anyone using
fmcdaq2.c for inspiration in an 8-lane design will get tripped up by the
default lane routing.

I'm not sure offhand, but there may also be power implications for
fmcdaq2 (without this patch, all 4 of the upper 4 logical lanes take
data from the 0th physical lane, which is actually active).

---
 fmcdaq2/fmcdaq2.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/fmcdaq2/fmcdaq2.c b/fmcdaq2/fmcdaq2.c
index 35ad30e..1c0a8c8 100644
--- a/fmcdaq2/fmcdaq2.c
+++ b/fmcdaq2/fmcdaq2.c
@@ -239,6 +239,8 @@ int fmcdaq2_reconfig(struct ad9144_init_param
*p_ad9144_param,

******************************************************************************/
 int main(void)
 {
+       int n;
+
        spi_init_param  ad9523_spi_param;
        spi_init_param  ad9144_spi_param;
        spi_init_param  ad9680_spi_param;
@@ -444,10 +446,8 @@ int main(void)
        ad9144_param.jesd204_subclass = 1;
        ad9144_param.jesd204_scrambling = 1;
        ad9144_param.jesd204_mode = 4;
-       ad9144_param.jesd204_lane_xbar[0] = 0;
-       ad9144_param.jesd204_lane_xbar[1] = 1;
-       ad9144_param.jesd204_lane_xbar[2] = 2;
-       ad9144_param.jesd204_lane_xbar[3] = 3;
+       for(n=0; n<ARRAY_SIZE(ad9144_param.jesd204_lane_xbar); n++)
+               ad9144_param.jesd204_lane_xbar[n] = n;

        ad9144_core.no_of_channels = 2;
        ad9144_core.resolution = 16;
--
2.20.1

gsmecher commented:
Anything I can do to move this merge forward? I see some formatting violations, but they're scattered throughout the file and not confined to my diffs. Thanks!

dbogdan commented:
Those formatting issues are present in the entire driver, not only in your additions, so we can fix everything through another commit.
However, these two commits are not signed, so please do that.

dbogdan commented on drivers/ad9144/ad9144.c:
```diff
@@ -492,6 +492,14 @@ int32_t ad9144_setup(struct ad9144_dev **device,
 	ad9144_spi_write(dev, REG_SYNC_CTRL, 0x81);	// sync-enable
 	ad9144_spi_write(dev, REG_SYNC_CTRL, 0xc1);	// sysref-armed
+	ad9144_spi_write(dev, REG_XBAR_LN_0_1,
+			init_param->jesd204_lane_xbar[0] | init_param->jesd204_lane_xbar[1]<<3);
```
Please use the SRC_LANE0(), SRC_LANE1() MACROs instead of direct shifting. Do it similar for the other lanes.

gsmecher commented:
@dbogdan, thanks for the prompt response! I have applied your suggestions, rebased the PR and added a signed-off-block. Please let me know if there's anything else I can do to help.

best,
Graeme

dbogdan commented:
@gsmecher, thanks for the patch.
